improve DiscriminantKind handling

lcnr opened the pull request:
Adds a lang item `discriminant_type` for the associated type `DiscriminantKind::Discriminant`.

Changes the discriminant of generators from `i32` to `u32`, which should not be observable to fix an
oversight where MIR was using `u32` and codegen and typeck used `i32`.

rust-highfive commented:
r? @matthewjasper

(rust_highfive has picked a reviewer for you, use r? to override)

nagisa commented on src/librustc_hir/lang_items.rs:
```diff
@@ -194,6 +194,8 @@ language_item_table! {
     CloneTraitLangItem,          "clone",              clone_trait,             Target::Trait;
     SyncTraitLangItem,           "sync",               sync_trait,              Target::Trait;
     DiscriminantKindTraitLangItem,"discriminant_kind", discriminant_kind_trait, Target::Trait;
+    DiscrKindDiscrLangItem,      "discr_kind_discr",   discr_kind_discr,        Target::AssocTy;
```
Maybe just `discriminant_type`?

nagisa reviewed (approved):
You mentioned that discriminant for `Generators` became `u32`. Can you elaborate "how" (in e.g. the commit message), given that the changes otherwise only involve removing the previously present `i32`… default.

Otheriwse r=me on implementation...

bors commented:
:umbrella: The latest upstream changes (presumably #74175) made this pull request unmergeable. Please resolve the merge conflicts.

lcnr committed:
improve DiscriminantKind handling

This now reuses `fn discriminant_ty` in project, removing
some code duplication. Doing so made me realize that
we previously had a disagreement about the discriminant
type of generators, with MIR using `u32` and codegen and
trait selection using `i32`.

We now always use `u32`.

lcnr commented:
Renamed the lang item to `discriminant_type` and added a note to the commit message explaining the change to the discriminant type of generators.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit fcf52c167f9280efe332557195998650a99bf437 has been approved by `nagisa`

<!-- @bors r=nagisa fcf52c167f9280efe332557195998650a99bf437 -->
<!-- homu: {"type":"Approved","sha":"fcf52c167f9280efe332557195998650a99bf437","approver":"nagisa"} -->
